mlflow/utils/autologging_utils.py

```diff
@@ -565,6 +579,37 @@ def safe_patch_function(*args, **kwargs):
         # Whether or not an exception was raised from within the original / underlying function
         # during the execution of patched code
         failed_during_original = False
+        # Whether or not to exclude auto-autologged content from content explicitly logged via
+        # `mlflow.start_run()`
+        exclusive = config.get("exclusive", False)
+
+        active_run = mlflow.active_run()
+
+        if active_run and exclusive and not _AutologgingSessionManager.active_session():
+            return original(*args, **kwargs)
+
+        # if active_run:
+        #   # there is an active run. could have been produced via user mlflow.start_run() OR a
+        #   # previously invoked, patched function.
+        #     if came from user (equivalent to "are we not in an autologging session?"):
+        #         if exclusive:
+        #             # dont patch the function; run original function. we dont want new content to be added
+        #             # to the run, or a child run to be created.
+        #             return original
+        #         else:
+        #             # carry on, patch the function, let the autologged content be added to that run
+        #             pass
+        #     else:
+        #         # must have come from us. patch function
+        #         pass
+        # else:
+        #     # need to start a run. patch function
+        #     pass
+
+        # Marker for all future function patches that will have to go through this process;
+        # now, when a run is active for those functions, they know it came from autologging and not from
+        # user, so they can go ahead and patch
+        _AutologgingSessionManager.start_session()
```

Comment:
Can we introduce a context manager that calls `start_session()` (or have `start_session` produce a context manager). This manager should call `end_session()` when it exits.

The rationale here is that, if for some reason the patch code terminates unexpectedly mid-way, we may never end the session. (See https://github.com/dbczumar/mlflow/pull/18/files#r540608305).

We'll also want to make sure to check for an active session before starting another (if one already exists, we should not start another).

Reply:
Sounds good!